src/calibre/utils/speedup.c

```diff
@@ -353,6 +357,41 @@ clean_xml_chars(PyObject *self, PyObject *text) {
     ans->length = j;
     return (PyObject*)ans;
 }
+#else
+static PyObject*
+clean_xml_chars(PyObject *self, PyObject *text) {
+    PyObject *result = NULL;
+    Py_UCS4 *ucs4_text = NULL;
+    Py_ssize_t src_i, target_i;
+    Py_UCS4 ch;
+
+    if (!PyUnicode_Check(text)) {
+        PyErr_SetString(PyExc_TypeError, "A unicode string is required");
+        return NULL;
+    }
+
+    ucs4_text = PyUnicode_AsUCS4Copy(text);
+    if (ucs4_text == NULL) return PyErr_NoMemory();
+
+    target_i = 0;
+    for (src_i = 0; src_i < PyUnicode_GET_LENGTH(text); src_i++) {
+        ch = ucs4_text[src_i];
+        if ((0x20 <= ch && ch <= 0xd7ff && ch != 0x7f) ||
+                ch == 9 || ch == 10 || ch == 13 ||
+                (0xe000 <= ch && ch <= 0xfffd) ||
+                (0xffff < ch && ch <= 0x10ffff)) {
+            // we can overwrite the same buffer that we're reading from because
+            // all our writes will occur at or before the current character
+            ucs4_text[target_i] = ch;
+            target_i += 1;
+        }
+    }
+    result = PyUnicode_FromKindAndData(
+        PyUnicode_4BYTE_KIND, ucs4_text, target_i);
```

Comment:
Do we really need to allocate the buffer twice? Also I dont like up-converting all strings to UCS 4 internally. It would be bad for performance since most XML data will be either ascii or ucs 2 at worst.

Reply:
> Do we really need to allocate the buffer twice?

Yes. PyUnicode_New() explicitly says that the objects created are not
resizeable. Therefore 2 allocations are needed:

1. the scratch space in which the filtered data is written
2. the PyObject that is returned

The python 2 code above digs into cpython implementation details (`ans->length
= j;` could be written as `PyUnicode_GET_SIZE(ans) = j;`).

There are a couple other options here:
 - do two passes over the array, first calculating the required length
 - ignore the docs and overwrite the PyUnicode length

> Also I dont like up-converting all strings to UCS 4 internally. It would be
> bad for performance since most XML data will be either ascii or ucs 2 at
> worst.

You're right, there is significant performance impact:

py3 gcc-4.8:

```console
$ python -m timeit -s  'import speedup; v = "a"*1000000000' -n 2 'speedup.clean_xml_chars(v)'
2 loops, best of 5: 2.81 sec per loop
```

py2:

```console
$ python -m timeit -s  'import speedup; v = u"a"*1000000000' -n 2 'speedup.clean_xml_chars(v)'
2 loops, best of 3: 1.46 sec per loop
```

However, most of this worse performance is inside `clean_xml`, and additional
copying is ~8% of the time. 

I'll see what I can do about this.

Reply:
The way I see it working is the following:

1) Basically write three versions of the function for ascii string, ucs2 and ucs4. The character validity checks in the case of ascii are also much simpler, which should be a small speedup

2) The vast majority of the time this function really does nothing, i.e.e there are no invalid chars in the XML. Therefore I think we should simply allocate a single buffer at the same size as the original string using PyMem_NEW(), fill it up and then create a unicode objet using PyUnicode_FromKindAndData